File: src/tools/orchestrate.ts
import { z } from "zod";
import type { AgentDefinition } from "../types.js";

const OrchestrateInputSchema = z.object({
  findings: z
    .string()
    .describe("JSON string of review findings from specialist agents"),
  context: z
    .string()
    .optional()
    .describe("Additional context about the review scope or priorities"),
});

export type OrchestrateInput = z.infer<typeof OrchestrateInputSchema>;

export { OrchestrateInputSchema };

export const orchestrateInputJsonSchema = {
  type: "object" as const,
  properties: {
    findings: {
      type: "string" as const,
      description: "JSON string of review findings from specialist agents",
    },
    context: {
      type: "string" as const,
      description:
        "Additional context about the review scope or priorities",
    },
  },
  required: ["findings"],
};

export function buildOrchestratePrompt(
  agent: AgentDefinition,
  findings: string,
  context?: string
): string {
  const parts: string[] = [];

  parts.push(agent.system_prompt);

  if (agent.checklist) {
    parts.push("\n## Process\n");
    parts.push(agent.checklist);
  }

  parts.push("\n## Specialist Findings\n");
  parts.push("```json");
  parts.push(findings);
  parts.push("```");

  if (context) {
    parts.push("\n## Additional Context\n");
    parts.push(context);
  }

  if (agent.output_format) {
    parts.push("\n## Required Output Format\n");
    parts.push(agent.output_format);
  }

  return parts.join("\n");
}

export async function executeOrchestrate(
  agent: AgentDefinition,
  input: OrchestrateInput
): Promise<{ content: string; isError: boolean }> {
  try {
    JSON.parse(input.findings);
  } catch (err) {
    return {
      content: `Invalid findings JSON: ${err instanceof Error ? err.message : String(err)}. Expected a JSON string from specialist agent output.`,
      isError: true,
    };
  }

  const prompt = buildOrchestratePrompt(agent, input.findings, input.context);
  return { content: prompt, isError: false };
}

File: rust/src/tools/orchestrate.rs
use crate::types::AgentDefinition;
use crate::tools::review::ToolResult;

fn build_orchestrate_prompt(
    agent: &AgentDefinition,
    findings: &str,
    context: Option<&str>,
) -> String {
    let mut parts = Vec::new();

    parts.push(agent.system_prompt.clone());

    if !agent.checklist.is_empty() {
        parts.push("\n## Process\n".to_string());
        parts.push(agent.checklist.clone());
    }

    parts.push("\n## Specialist Findings\n".to_string());
    parts.push(findings.to_string());

    if let Some(ctx) = context {
        parts.push("\n## Additional Context\n".to_string());
        parts.push(ctx.to_string());
    }

    if !agent.output_format.is_empty() {
        parts.push("\n## Required Output Format\n".to_string());
        parts.push(agent.output_format.clone());
    }

    parts.join("\n")
}

pub fn execute_orchestrate(
    agent: &AgentDefinition,
    findings: &str,
    context: Option<&str>,
) -> ToolResult {
    if findings.trim().is_empty() {
        return ToolResult {
            content: "Empty findings string. Expected structured Markdown from specialist agent output.".to_string(),
            is_error: true,
        };
    }

    let prompt = build_orchestrate_prompt(agent, findings, context);
    ToolResult {
        content: prompt,
        is_error: false,
    }
}
